yjit/src/codegen.rs

```diff
@@ -7844,6 +7848,39 @@ fn gen_getblockparamproxy(
 
         let top = asm.stack_push(Type::BlockParamProxy);
         asm.mov(top, Opnd::const_ptr(unsafe { rb_block_param_proxy }.as_ptr()));
+    } else if unsafe { rb_obj_is_proc(comptime_handler) }.test() {
+        // The block parameter is a Proc
+        c_callable! {
+            // We can't hold values across C calls due to a backend limitation,
+            // so we'll use this thin wrapper around rb_obj_is_proc().
+            fn is_proc(object: VALUE) -> VALUE {
+                if unsafe { rb_obj_is_proc(object) }.test() {
+                    // VM_BH_TO_PROC() is the identify function.
+                    object
+                } else {
+                    Qfalse
+                }
+            }
+        }
+
+        // Simple predicate, no need to jit_prepare_routine_call()
+        let proc_or_false = asm.ccall(is_proc as _, vec![block_handler]);
+
+        // Guard for proc
+        asm.cmp(proc_or_false, Qfalse.into());
+        jit_chain_guard(
+            JCC_JE,
+            jit,
+            asm,
+            ocb,
+            SEND_MAX_DEPTH,
+            Counter::gbpp_block_handler_not_proc,
+        );
+
+        let top = asm.stack_push(Type::Unknown);
+        asm.mov(top, proc_or_false);
+    } else {
+        unreachable!("absurd given initial filtering");
```

Comment:
I'm tempted to put `gen_counter_incr(asm, Counter::gbpp_unsupported_type);` and `return None;` here to simplify the logic (and possibly optimize away the wasted instructions by rewinding `Assembler` in the future), but do you think this case is frequent enough?

Reply:
Move the check here, you mean? I think that makes sense. The symbol case seems uncommon enough.

Reply:
Oh, that ends up triggering a different issue: `attempted to lower an Opnd::InsnOut` probably because the `block_handler` ends up unused :/ so it doesn't seem enticing anymore.

Reply:
Ah, that issue. Never mind then.